### ISSUE TITLE
False positive try-except-raise

### ISSUE THREAD

**Issue by sam-s** [PyCQA/pylint#3083]
### Steps to reproduce
When you need to [catch an exception in the `for` loop iterator](https://stackoverflow.com/q/13653783/850781), you write something like this:
```python
def wrapper(gen):
  while True:
    try:
      yield next(gen)
    except StopIteration:
      raise
    except Exception as e:
      print(e) # or whatever kind of logging you want
 ```
In this case, the immediate `raise` after `except` is _correct_ because there is another `except` after that.

### Current behavior

```
foo.py:109: warning (W0706, try-except-raise, cursor_wrapper) The except handler raises immediately
```

### Expected behavior

No diagnostics

### pylint --version output
```
pylint 2.3.1
astroid 2.2.5
Python 3.7.4 (default, Jul  9 2019, 18:13:23) 
[Clang 10.0.1 (clang-1001.0.46.4)]
```

**Comment by PCManticore** [comment]
Thanks for creating the issue, this makes sense and it's definitely a bug.

**Comment by PCManticore** [comment]
Gave this another go and cannot reproduce it. We also have code around these except handlers being followed by generic ones, so not sure how this issue was triggered for you. Do let me know if you had a different example for which you can reproduce the issue consistently using the latest `pylint`.

**Comment by PCManticore** [comment]
Still cannot reproduce with 3.7.2, pylint 2.3.1 and astroid 2.2.5.

**Comment by sam-s** [comment]
```
pylint 2.4.2
astroid 2.3.1
Python 3.7.4 (default, Sep  7 2019, 18:27:02) 
[Clang 10.0.1 (clang-1001.0.46.4)]
```

**Comment by SamyCookie** [comment]
similar issue with this use case:
```python3
        try:
            raise Exception(42)
        except Exception:
            raise
        finally:
            print(42)
```
with
```
pylint 2.4.3
astroid 2.3.2
Python 3.5.3 (default, Sep 27 2018, 17:25:39) 
[GCC 6.3.0 20170516]
```

**Comment by GoldsteinE** [comment]
Same for me:
```python
import asyncio

try:
    _ = 1
except asyncio.CancelledError:
    raise
except Exception as err:
    raise ValueError
```
Interesting that if `asyncio.CancelledError` is changed to e.g. `TypeError`, pylint shows no warning.
```
pylint 2.4.4
astroid 2.3.3
Python 3.8.1 (default, Dec 21 2019, 20:57:38)
[GCC 9.2.0]
```

**Comment by TheKevJames** [comment]
@GoldsteinE could you expand on "asyncio.CancelledError is inherited from BaseException and will not be caught by except Exception: branch"? I seem to be getting different behavior than I'd expect:

```
» python3
Python 3.7.10 (default, Mar 17 2021, 21:29:25)
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
>>> import asyncio
>>> try:
...     raise asyncio.CancelledError()
... except Exception:
...     print('exc!')
...     raise
...
exc!
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
concurrent.futures._base.CancelledError
>>> try:
...     raise asyncio.CancelledError()
... except asyncio.CancelledError:
...     print('not an exc!')
... except Exception:
...     print('exc!')
...     raise
...
not an exc!
```

This in turn implies to me that they would be many scenarios where folks would want to `try-except-raise` for asyncio.CancelledErrors. I indeed am continuing to get this false positive with pylint v2.8.2 and the following reproduction code:
```python
try:
    _ = 0
except asyncio.CancelledError:
    raise
except Exception:
    print('Got an exception!')
```

**Comment by GoldsteinE** [comment]
It's a Python version issue. `CancelledError` stopped being a subclass of `Exception` since 3.8.

**Comment by jacobtylerwalls** [comment]
Greetings @SamyCookie I think the [example](https://github.com/PyCQA/pylint/issues/3083#issuecomment-549919190) you posted is not a bug. If except immediately raises without a further except clause, there's no filtering of the exception types, and Pylint's behavior is correct.

The update from the author of https://github.com/PyCQA/pylint/issues/3083#issuecomment-572545715 also says not a bug.

And https://github.com/PyCQA/pylint/issues/3083#issuecomment-540637016 is not a bug, since the `bson` exception is not a superset of `StopIteration`, so catching `StopIteration` and re-raising it is pointless (or only for readability, in which case the warning can be suppressed).

So we may be able to close.

**Comment by SamyCookie** [comment]
Thanks @jacobtylerwalls, I don't know why I saw a bug there, so okay for closing this !

**Comment by sam-s** [comment]
Wait, why did you close this?!
This behavior is still present, and it is incorrect: it warns on a perfectly valid code.

**Comment by jacobtylerwalls** [comment]
`StopIteration` is not a subclass of `bson.errors.InvalidBSON`, so raising it immediately after catching it is an anti-pattern. If you find it opinionated, that's a reason for disabling it, but it behaves in this case exactly as advertised.

**Comment by Pierre-Sassoulas** [comment]
And the case where the exception caught later on is a subclass of the first one is handled correctly:
```python
# pylint: disable=missing-docstring

def function(gen):
    try:
        yield next(gen)
    except StopIteration:
        raise
    except Exception as exc:
        print(exc) # or whatever kind of logging you want

```

``pylint b.py --enable=all --enable-all-extensions --disable=I``:
```
************* Module b
b.py:8:11: W0703: Catching too general exception Exception (broad-except)

------------------------------------------------------------------
Your code has been rated at 8.57/10 (previous run: 8.57/10, +0.00)
```

**Comment by sam-s** [comment]
> `StopIteration` is not a subclass of `bson.errors.InvalidBSON`, so raising it immediately after catching it is an anti-pattern.

I don't follow, sorry. How do I write the code to avoid the warning?

**Comment by chaoflow** [comment]
@sam-s In your example posted [here](https://github.com/PyCQA/pylint/issues/3083#issuecomment-540637016), why do you see the need for ``except StopIteration``? It would not be caught by any following ``except`` clause and therefore does not need to be specially handled. The special handling is only needed if the special exception is a *subclass* of a more generic exception being handled, like in @Pierre-Sassoulas [example](https://github.com/PyCQA/pylint/issues/3083#issuecomment-979437681) where ``StopIteration`` is a subclass of ``Exception``.

**Comment by Pierre-Sassoulas** [comment]
``StopIteration is not a subclass of bson.errors.InvalidBSON`` means that catching ``StopIteration`` and reraising immediately does nothing. So you can remove the code that catch and reraise:

```python
import sys

import bson

class Pull():
    "Pull data from MongoDB"
    def __init__(self):
        self.cursor_bson_errors = []

    def bson_cursor_wrapper(self, cursor):
        """Catch bson errors and report them in details.
        https://stackoverflow.com/q/13653783/850781
        https://stackoverflow.com/q/46933995/850781"""
        while True:
            try:
                yield next(cursor)
            except bson.errors.InvalidBSON as exn:
                _exc_type, exc_value, exc_traceback = sys.exc_info()
                assert exn == exc_value
                last_tb = exc_traceback
                while last_tb.tb_next:
                    last_tb = last_tb.tb_next
                # File "/.../bson/__init__.py", line 1098, in _decode_all_selective
                self.cursor_bson_errors.append(len(last_tb.tb_frame.f_locals["data"]))
```

For this snippet StopIteration is a subclass of Exception
```python
def function(gen):
    try:
        yield next(gen)
    except StopIteration:
        raise
    except Exception as exc:
        print(exc) # or whatever kind of logging you want
```

Which means that catching and re-raising ``StopITeration`` prevents the function from being printed instead of raised. It's a specific treatment you apply to ``StopIteration``. And it's not equivalent to:

```python
def function(gen):
    try:
        yield next(gen)
    except Exception as exc:
        print(exc) # or whatever kind of logging you want
```
